homeassistant/components/amcrest/camera.py

```diff
@@ -112,28 +118,58 @@ def __init__(self, name, device, ffmpeg):
         self._motion_recording_enabled = None
         self._color_bw = None
         self._rtsp_url = None
-        self._snapshot_lock = asyncio.Lock()
+        self._snapshot_task = None
         self._unsub_dispatcher = []
         self._update_succeeded = False
 
-    async def async_camera_image(self):
-        """Return a still image response from the camera."""
+    def _check_snapshot_ok(self):
         available = self.available
         if not available or not self.is_on:
             _LOGGER.warning(
                 "Attempt to take snaphot when %s camera is %s",
                 self.name,
                 "offline" if not available else "off",
             )
+            raise _CannotSnapshot
+
+    async def _async_get_image(self):
+        try:
+            # Send the request to snap a picture and return raw jpg data
+            # Snapshot command needs a much longer read timeout than other commands.
+            return await self.hass.async_add_executor_job(
+                partial(
+                    self._api.snapshot,
+                    timeout=(COMM_TIMEOUT, SNAPSHOT_TIMEOUT),
+                    stream=False,
+                )
+            )
+        except AmcrestError as error:
+            log_update_error(_LOGGER, "get image from", self.name, "camera", error)
+            return None
+        finally:
+            self._snapshot_task = None
+
+    async def async_camera_image(self):
+        """Return a still image response from the camera."""
+        _LOGGER.debug("Take snapshot from %s", self._name)
+        try:
+            # Amcrest cameras only support one snapshot command at a time.
+            # Hence need to wait if a previous snapshot has not yet finished.
+            # Also need to check that camera is online and turned on before each wait
+            # and before initiating shapshot.
+            while self._snapshot_task:
```

Comment:
I think it would be more clear to use a asyncio lock to control the access to the of snapshot.

Why did we change this?

Reply:
Basically for the reason explained in the comment. The call to `async_camera_image` can get canceled, and when that happens, and it does because sometimes the camera takes too long to complete the snapshot command, there is no context to catch an exception, which can also happen in this scenario. But more importantly, if the call to `async_camera_image` is canceled, then the lock will automatically be released, and hence will not do its job to prevent the next call from trying to start a new snapshot command when the last one hasn't finished, which leads to communication errors with the camera. Using a task that can't get canceled solves that problem. This one change alone almost entirely eliminated all the warnings and errors I used to get with the way it was (i.e., just using a lock.)